app/services/migration_service.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class MigrationService
+  attr_reader :archive_path, :new_user_name
+  delegate :errors, :warnings, to: :archive_validator
+
+  def initialize(args)
+    @archive_path = args[:archive_path]
+    @new_user_name = args[:new_user_name]
+  end
+
+  def validate_archive
+    archive_validator.validate
+    raise ArchiveValidationFailed, archive_validator.errors if archive_validator.errors.any?
+  end
+
+  def perform!
+    create_user
+    import_archive
+    run_migration
+  end
+
+  private
+
+  def run_migration
+    account_migration.save
+    account_migration.perform!
+  end
+
+  def import_archive
+    archive_importer.import
+  end
+
+  def create_user
+    archive_importer.create_user(new_user_name)
+  end
+
+  def account_migration
+    @account_migration ||= AccountMigration.new(
+      old_person:      Person.by_account_identifier(archive_importer.archive_author_diaspora_id),
```

Comment:
We discussed that we shouldn't store the `AccountMigration`, but we actually have to send it to other nodes, therefore we should perform it, because by current design dispatching is a part of performing.

Also `account_migration` method won't return nil when `old_account` doesn't exist. Instead it will return a non-persited object:
```ruby
irb(main):001:0> AccountMigration.new
=> #<AccountMigration id: nil, old_person_id: nil, new_person_id: nil>
```

So the idea is that while validation will fail during the attempt to save the object, everything else will run the same way, just as we have discussed: no matter if the pod is aware of the old account or not, the procedure of migration/import is the same.

`perform` method is:
```ruby
  def perform!
    raise "already performed" if performed?
    validate_sender if locally_initiated?
    identity_transplantation if old_person
    dispatch if locally_initiated?
    dispatch_contacts if remotely_initiated?
  end
```

And if `old_person` is unavailable it will only run dispatching, so this pretty much work as expected.

Reply:
I can see though that I didn't test the case and currently `perform!` actually fails to run when `old_person` is nil. I'll work on a fix.

Reply:
So I added [the commit](https://github.com/diaspora/diaspora/pull/7660/commits/259e5522876f46e8e020c34a0172ffc6be9ab812) which fixes the work flow for the case of missing and unfetchable user.

Reply:
> So the idea is that while validation will fail during the attempt to save the object, everything else will run the same way, just as we have discussed: no matter if the pod is aware of the old account or not, the procedure of migration/import is the same.

Oh, yes, that makes sense to me and it looks good now. :+1: